client/main/infobox.js

```diff
@@ -28,39 +29,151 @@ class InfoBox extends HTMLElement {
-        this._button = this._body.querySelector('button');
+        this._button = this._body.querySelector('.button-box button');
+        this._close = this._body.querySelector('.close');
+
+        this._close.addEventListener('click', () => this.hide());
+        this._host.addEventListener('click', (event) => this._backgroundClicked(event));
+        this._host.addEventListener('keydown', (event) => this._onKeyDown(event));
 
         this._button.addEventListener('click', () => this.clicked());
         window.addEventListener('message', (event) => this.onMessage(event));
     }
 
-    setup(info) {
-        this._host.setAttribute('title', info.title || '');
-        this._host.setAttribute('message', info.message || '');
-        this._host.setAttribute('message-src', info['message-src'] || '');
-        this._host.setAttribute('status', info.status || '');
+    setup(info, params) {
+        if (params === undefined)
+            params = { };
+
+        if (this._displayInfo)
+            this.hide();
+
+        this._visible = true;
+        this._processEnterKey = false;
+
+        this._displayInfo = info;
+        if (info['message-src']) {
+            if (params.cancelable === undefined)
+                params.cancelable = false;
+
+            this._local.classList.remove('external');
+            this._body.style.height = `300px`;
+            this._body.style.width = `300px`;
+            this._host.setAttribute('message-src', info['message-src'] || '');
+            this._host.style.display = null;
+            keyboardJS.pause();
+        }
+        else {
+            if (params.cancelable === undefined)
+                params.cancelable = true;
+
+            this._local.classList.add('external');
+            let show = true;
+            if (status.message || info.title) {
+                this._host.setAttribute('title', info.title || '');
+                this._host.setAttribute('message', info.message || '');
+                this._host.setAttribute('status', info.status || '');
+                this._processEnterKey = true;
+            }
+            else if (this._isElement(info)) {
+                this.appendChild(info);
+            }
+            else
+                show = false;
+
+            if (show) {
+                keyboardJS.pause();
+                this._host.style.display = null;
+                setTimeout(() => {
+                    this._body.style.opacity = 1;
+                    this._host.style.opacity = 1;
+                }, 10);
+            }
+        }
+
+        this._cancelable = params.cancelable;
+
+        if (this._cancelable) {
+            this._close.style.display = 'block';
+        }
+    }
+
+    _isElement(item){
+      return ( typeof HTMLElement === "object" ? item instanceof HTMLElement : //DOM2
+            item && typeof item === "object" && item !== null && item.nodeType === 1 && typeof item.nodeName==="string");
+    }
+
+    hide() {
```

Comment:
this approach is probably fine, but a weakness of it is that it requires the `hide()` function to be called explicitly. i.e. if someone simply removes the infobox from the DOM in the normal way, then this will leave gunk behind (i.e. it will still be eating keyboard events). i'm happy to merge this as-is, but it might be worth thinking about plumbing into those DOM events as a more robust approach.

Reply:
I think this is in on the phone convo. Or skype